modules/dhcp_native_ms/dhcp_native_ms_main.rb

```diff
@@ -77,8 +78,27 @@ def unused_ip(subnet_address, mac_address, from_address, to_address)
       if dhcpsapi.api_level == DhcpsApi::Server::DHCPS_WIN2008_API || dhcpsapi.api_level == DhcpsApi::Server::DHCPS_NONE
         raise Proxy::DHCP::NotImplemented.new("DhcpsApi::Server#get_free_ip_address is not available on Windows Server 2008 and earlier versions.")
       end
-
-      return dhcpsapi.get_free_ip_address(subnet_address, from_address, to_address).first
+      
+      logger.debug "Searching for free IP in subnet #{subnet_address}"
+      while true
+        ip = dhcpsapi.get_free_ip_address(subnet_address, from_address, to_address).first
+        if ip.nil? || ip.empty?
+          logger.warn "No free IP returned by DHCP for subnet #{subnet_address}"
+          return nil
+        end
+        if system("ping -n 1 -w 1000 #{ip} > NUL")
+          logger.debug "Found a pingable IP address which does not have a DHCP record: #{ip}"
+        else
+          logger.info "Found free IP #{ip}"
+          return ip
+        end
+        break if ip == to_address
+        found_ip = IPAddr.new(ip, Socket::AF_INET)
+        next_ip = IPAddr.new(found_ip.to_i + 1, Socket::AF_INET)
```

Comment:
Please call #get_free_ip_address but with an updated range instead: otherwise this can end up selecting addresses from the exclusion range, or iterating over already allocated addresses, which is slow.

Reply:
Sorry I don't get you. The from_address is updated after each get_free_ip_address call with the last returned free ip + 1 (in case it is pingable). I considered that get_free_ip_address never returns an already allocated address or an address part of an exclusion range. Am I wrong? Unfortunately I can't test with exclusion range because of http://projects.theforeman.org/issues/21237

Reply:
correction: I was able to test with a small exclusion range (not full range) and the get_free_ip_address correctly handle that. It does not give an IP part of an exclusion range.

Reply:
Apologies, I missed that your implementation already does what I suggested.